apps/notifications/src/interfaces/rest/decorators/http-validate/http-validate.decorator.ts

```diff
@@ -1,19 +1,81 @@
 import { applyDecorators, UsePipes } from "@nestjs/common";
-import { ApiCreatedResponse } from "@nestjs/swagger";
+import { ApiResponse } from "@nestjs/swagger";
 import type { ZodDto } from "nestjs-zod";
+import { ZodValidationException } from "nestjs-zod";
+import { createZodDto } from "nestjs-zod";
 import { ZodSerializerDto, ZodValidationPipe } from "nestjs-zod";
 import type { ZodTypeDef } from "zod";
+import { z } from "zod";
 
-export function ValidateHttp(options: { response?: ZodDto<any, ZodTypeDef, any> }) {
-  const decorators = [
-    UsePipes(ZodValidationPipe),
-    ApiCreatedResponse({
-      type: options.response
-    })
-  ];
+const notFoundErrorResponseSchema = z.object({
+  statusCode: z.literal(404),
+  message: z.string()
+});
 
-  if (options?.response) {
-    decorators.push(ZodSerializerDto(options.response));
+export class NotFoundErrorResponse extends createZodDto(notFoundErrorResponseSchema) {}
+
+const internalServerErrorResponseSchema = z.object({
+  statusCode: z.literal(500),
+  message: z.string()
+});
+
+export class InternalServerErrorResponse extends createZodDto(internalServerErrorResponseSchema) {}
+
+const validationErrorResponseSchema = z.object({
+  statusCode: z.literal(400),
+  message: z.string(),
+  errors: z.object({
+    issues: z.array(z.object({}))
+  })
+});
+
+export class ValidationErrorResponse extends createZodDto(validationErrorResponseSchema) {}
+ZodValidationException;
+type ResponseDefinitionOptions = Record<
+  number,
+  {
+    description?: string;
+    schema: ZodDto<any, ZodTypeDef, any>;
+  }
+>;
+
+export function ValidateHttp(options: ResponseDefinitionOptions) {
+  const successSchema = (options[200] || options[201])?.schema;
+
+  const decorators = [UsePipes(ZodValidationPipe)];
+
+  for (const [statusCode, def] of Object.entries(options)) {
+    decorators.push(
+      ApiResponse({
+        status: Number(statusCode),
+        type: def.schema,
+        description: def.description
+      })
+    );
+  }
+
+  if (!options[500]) {
+    decorators.push(
+      ApiResponse({
+        status: 500,
+        type: InternalServerErrorResponse,
+        description: "Internal server error, should probably be reported"
```

Comment:
Is this shown to the client?

Reply:
Yea, it's the same logic as with all the other stuff. If anything unhandled happens it's this schema.

Reply:
Ah, actually this is not on the api response if that's what you're asking. It's for openapi/swagger